montblanc/impl/rime/tensorflow/rime_ops/zernike_op_gpu.cuh

```diff
@@ -149,33 +164,61 @@ __global__ void zernike_dde_zernike(
     int ant = blockIdx.y * blockDim.y + threadIdx.y;
     int time = blockIdx.z * blockDim.z + threadIdx.z;
 
-  int i = threadIdx.z * 4 + (threadIdx.x & 0x03);
+    if(corr >= 4 || chan >= nchan || ant >= na || time >= ntime) return;
 
-    if(i < npoly){  
-        shared.zernike_coeff[threadIdx.y][threadIdx.x >> 2][i] = in_coeffs[((ant * nchan + chan) * npoly + i) * 4 + corr];
-        shared.zernike_noll_indices[threadIdx.y][threadIdx.x >> 2][i] = in_noll_index[((ant * nchan + chan) * npoly + i) * 4 + corr];
+    if(threadIdx.z < npoly)
+    {
+        shared.zernike_coeff[threadIdx.y][threadIdx.x >> 2][threadIdx.z] = in_coeffs[
+            ((ant * nchan + chan) * npoly + threadIdx.z) * 4 + corr];
+        shared.zernike_noll_indices[threadIdx.y][threadIdx.x >> 2][threadIdx.z] = in_noll_index[
+            ((ant * nchan + chan) * npoly + threadIdx.z) * 4 + corr];
```

Comment:
Ah I can see the issue here -- I didn't forsee this. This requires the TIME dimension to be intimately related to the polynomial dimension which makes things awkward.

I'm sorry if this is frustrating, but could you change the above load of polynomials into shared memory to loop over polynomial again while retaining the  per ant/chan thread load. Then, time threads can be redistributed back over antenna and corrchan? In particular corrchan works best with at least 32 threads (its the size of a CUDA warp).